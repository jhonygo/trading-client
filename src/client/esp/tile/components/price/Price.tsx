import React from 'react';
import classNames from 'classnames';
import { TradeRequest } from '../../model/tradeRequest';
import * as style from './style.css';
import { Movements } from '../../model/priceTick';
import styled from 'styled-components';

export const TileStaleStyle = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  border-radius: 3px;
  padding: 1.25rem;
  line-height: 1.5;
  z-index: 5;
  opacity: 0.7;
  pointer-events: none;
  cursor: not-allowed;
  background-color: #3d4853;
`;

export namespace Price {
  export interface Props {
    symbol: string;
    side: string;
    price: number;
    notional: number;
    execute: (tradeRequest: TradeRequest) => void;
    executing: boolean;
    url: string;
    movement?: string;
    isStale: boolean;
  }

  export interface State {
    first: string;
    bigFigures: string;
    tenthOfPips: string;
  }
}

export class Price extends React.PureComponent<Price.Props, Price.State> {
  constructor(props: Price.Props, context: any) {
    super(props, context);
    this.state = {
      first: '0.00',
      bigFigures: '00',
      tenthOfPips: '0'
    };
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  handleOnClick() {
    const tradeRequest: TradeRequest = {
      id: 1,
      symbol: this.props.symbol,
      broker: 'WEB',
      price: this.props.price,
      side: this.props.side,
      quantity: this.props.notional,
      url: this.props.url
    };
    this.props.execute(tradeRequest);
  }

  componentWillReceiveProps(newProps) {
    this.setState(this.extractPrice(newProps.price));
  }

  extractPrice(price) {
    const priceStr = price.toString();
    const first = priceStr.substr(0, 4);
    const bigFigures = priceStr.substr(4, 2).padEnd(2, '00');
    const tenthOfPips = priceStr.substr(6) || '0';
    return {
      first: first,
      bigFigures: bigFigures,
      tenthOfPips: tenthOfPips,
      movement: price.movement
    };
  }

  render() {
    const { side, price, executing, movement, isStale } = this.props;
    const { first, bigFigures, tenthOfPips } = this.state;

    const classes = classNames({
      [style.oneWayPrice]: true,
      [style.buy]: side === 'buy',
      [style.sell]: side === 'sell',
      'card-box': true,
      [style.nonTradeable]: price === 0,
      [style.executing]: executing === true
    });

    const visible = {
      visibility: 'visible'
    } as React.CSSProperties;

    const hidden = {
      visibility: 'hidden'
    } as React.CSSProperties;

    const upStyle =
      movement === Movements.Up ? visible : (hidden as React.CSSProperties);

    const downStyle =
      movement === Movements.Down ? visible : (hidden as React.CSSProperties);

    return (
      <div
        className="col-md-5"
        id="request-execution"
        onClick={this.handleOnClick}
      >
        <div className={`card ` + classes}>
          {isStale && (
            <TileStaleStyle>
              <h5 className="card-title">{side}</h5>
              <div className="text-white bg-secondary">Stale</div>
            </TileStaleStyle>
          )}

          <div className="card-body">
            <h5 className="card-title">{side}</h5>
            <div className={style.spreadUp} style={upStyle} />
            <div>
              <span>
                <span className={style.priceValue}>{first}</span>
                <span className={style.bigFigures}>{bigFigures}</span>
                <span className={style.tenthOfPips}>{tenthOfPips}</span>
              </span>
            </div>
            <div className={style.spreadDown} style={downStyle} />
          </div>
        </div>
      </div>
    );
  }
}
