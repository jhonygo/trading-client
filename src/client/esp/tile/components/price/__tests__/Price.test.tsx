import React from 'react';
import { shallow } from 'enzyme';
import { Price } from '../Price';

describe('Price', () => {
  it('should extract price value', () => {
    // Given
    const price = shallow(
      <Price
        symbol="EURUSD"
        price={1.37283}
        side="buy"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
      />
    );

    // When
    price.setProps({ price: 1.18302 });

    // Then
    expect(price.state().first).toEqual('1.18');
    expect(price.state().bigFigures).toEqual('30');
    expect(price.state().tenthOfPips).toEqual('2');
  });

  it('should default price decimal to match symbol precision', () => {
    // Given
    const price = shallow(
      <Price
        symbol="EURUSD"
        price={1.37283}
        side="buy"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
      />
    );

    // When
    price.setProps({ price: 1.18 });

    // Then
    expect(price.state().first).toEqual('1.18');
    expect(price.state().bigFigures).toEqual('00');
    expect(price.state().tenthOfPips).toEqual('0');
  });

  it('should request execution', () => {
    // Given
    const execute = jest.fn();
    const price = shallow(
      <Price
        symbol="USDGBP"
        price={1.18302}
        side="buy"
        notional={10000000}
        execute={execute}
        executing={false}
      />
    );

    // When
    price.find('#request-execution').simulate('click');

    // Then
    expect(execute).toHaveBeenCalledWith({
      id: 1,
      symbol: 'USDGBP',
      broker: 'WEB',
      price: 1.18302,
      side: 'buy',
      quantity: 10000000
    });
  });

  it('should render buy price update', () => {
    // Given
    const price = shallow(
      <Price
        symbol="USDGBP"
        price={1.18302}
        side="buy"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
      />
    );

    // price
    price.setProps({ price: 1.43841 });

    // Then
    expect(price).toMatchSnapshot();
  });

  it('should render sell price update', () => {
    // Given
    const price = shallow(
      <Price
        symbol="USDGBP"
        price={1.18302}
        side="sell"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
      />
    );

    // When
    price.setProps({ price: 1.43841 });

    // Then
    expect(price).toMatchSnapshot();
  });
});
