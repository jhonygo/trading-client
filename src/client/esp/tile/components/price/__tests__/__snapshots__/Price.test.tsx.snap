// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price should render buy price update 1`] = `ShallowWrapper {}`;

exports[`Price should render sell price update 1`] = `ShallowWrapper {}`;
